### PR TITLE
fix replies compatibility

### LINKED ISSUE #6903: Replies are shown as ordinary messages if they are sent from current nightly and quoted on release build or when reply to old message sent before upgrade and both contacts are on release build
# Case 1: Replies are shown as ordinary messages if they are sent from current nightly and quoted on release build
### Description

*Type*: Bug
*Summary*: when user has current nightly installed and replies to message, sent from release build, user with release build will see an ordinary message and vice versa.

#### Expected behavior
can see reply message on both A and B

#### Actual behavior
ordinary message on device A and reply message from device B
deblkdjsf

### Reproduction

**Prerequisites:** 
1. device A has installed  0.9.31 version (released) 
2. device B has installed  latest nightly 26/11/2018

- Open Status on device A
- Log in
- Join any channel (i.e. #test)
- Open Status on device A
- Log in
- Join same channel (i.e. #test)
- Send a message from device A
- Reply to message from device B
- Check public chat on device A and B

### Additional Information
* Status version: nightly 26/11/2018
* Operating System: all

# Case 2:  No quotation is shown in the reply message on recipient side if reply to message sent before upgrade

### Description

[comment]: # (Feature or Bug? i.e Type: Bug)
*Type*: Bug

[comment]: # (Describe the feature you would like, or briefly summarise the bug and what you did, what you expected to happen, and what actually happens. Sections below)
*Summary*: Recipient gets no quotation in the reply message if reply to message sent before upgrade from 0.9.31 to 0.9.32. On sender side quotation is shown as expected. Both recipient and sender are on 0.9.32. Note that if reply to message sent after upgrade is done on both sides then quotation is shown as expected for both sender and recipient. So, issue is only on recipient side for replies to "old" messages sent while being on 0.9.31.

Recipient (no quotation is shown):

Sender (shows quotation as expected)
screenshot_20181129-160441

#### Expected behavior
[comment]: # (Describe what you expected to happen.)
Quotation is shown to recipient, sender and other chat participants for replies to messages sent on 0.9.31 and on 0.9.32 
#### Actual behavior
[comment]: # (Describe what actually happened.)
Quotation is not shown to recipient if reply to message sent before upgrade
### Reproduction
[comment]: # (Describe how we can replicate the bug step by step.)

1. sender and recipient are on 0.9.31 and they are in contacts of each other
2. both send text message in 1:1 chat
3. both upgrade to 0.9.32
4. sender replies to message from recipient in 1:1 chat
5. check how reply is shown on recipient side

### Additional Information
[comment]: # (Please do your best to fill this out.)
* Status version: from 0.9.31 release to build 7 of 0.9.32 release
* Operating System: Android and iOS

### PR DESCRIPTION
**Pull request by rasom**
fix #6903 
1. `old-message-id` field (indexed) was introduced in `message` entity and is calculated as `message-id` was calculated in `0.9.31`
```clojure
(defn old-message-id
  [message]
  (sha3 (pr-str message)))
```
2. When a reply message is sent from the PR version of app both `response-to` and `response-to-new` (probably better to call it `v2/response-to`) fields are sent as a part of `message`'s `content` field, so that it can be recognized by `0.9.31`.
3. When PR version of app receives reply from `0.9.31` we check whether message's `content` contains `response-to` but doesn't contain `response-to-new`, and if so we check whether DB contains message with `old-message-id=response-to`. If such message has been found we assoc `response-to-new` to content.
4. If message from DB contains only `response-to` but not `response-to-new` attempt to fetch the message by `old-message-id` is done (can be also handled by migration)

status: ready

### CODE REVIEW

**Comment by asemiankevich** [comment]
@rasom would you mind to rebase? e2e is ending again with `sha256` but i remember @jakubgs has fixed that on Friday

**Comment by asemiankevich** [comment]
@rasom case 2 from https://github.com/status-im/status-react/issues/6903 is still actual and i can reproduce it the way it is described:

1. sender and recipient are on 0.9.31 and they are in contacts of each other
2. both send text message in 1:1 chat
3. both upgrade to 0.9.32
4. sender replies to message from recipient in 1:1 chat
5. check how reply is shown on recipient side

NOTE: case 1 from the issue above is fixed.

cc @churik @annadanchenko

**Comment by rasom** [comment]
@asemiankevich `case 2` should be fixed now, branch was rebased

**Comment by asemiankevich** [comment]
@rasom thanks :) would be nice to have some rewies there :)

**Comment by asemiankevich** [comment]
so I dont see none of issues described in this PR build. @churik please confirm too (just to double check)

**Comment by cammellos** [comment]
@rasom could you add a description on the PR of what the logic is please?

Also, just to make sure, this does not change the id as described in https://github.com/status-im/status-react/issues/6956 , so if we push this out without, we probably want to make sure there's a clean upgrade path?

**Comment by rasom** [comment]
yes sure @cammellos, working on the description

> this does not change the id as described in #6956 (comment)

this pr doesn't change the way how `message-id` is calculated, right

**Comment by rasom** [comment]
@cammellos see description, idk if that's enough

**Comment by cammellos** [comment]
@rasom thanks looks good.
In terms of upgrade path, this is non breaking, if we decided to go for https://github.com/status-im/status-react/issues/6956#issuecomment-443622618 , is there a clean way not-to break compatibility? or we are ok breaking when/if we implement that? I think it would be good to figure it out before we push this out

**Comment by rasom** [comment]
>  I think it would be good to figure it out before we push this out

This is a separate thing, but yes, i would prefer to fix `message-id` before 0.9.32 release. Otherwise, we will have compatibility issues next time as well. I don't think we can guaranty anything in nightlies, it's just too much overhead imo.

**Comment by churik** [comment]
Both cases are not reproducible for me.
Case 1 checked on IOS 12.1, Iphone XS (release build) and Mac OS Mojave (PR build).
Case 2 checked on IOS 11.4, Iphone 7 and Android 8.0, Google Pixel (Genymotion) for public and 1-1 chat.

**Comment by asemiankevich** [comment]
@yenda @mandrigin review this please. @cammellos @rasom are you planning to create separate PR to fix the remaining stuff? we are planning this release schedule

**Comment by cammellos** [comment]
On my side, before including this I would like to answer this questions:

1) If we include it, and then change id scheme, is there a clear path to upgrade id without breaking?
2) If the answer of point 1 is `no`, are we ok with breaking compatibility for replies/seen etc? Otherwise cherry-pick
3) If the answer of point 2 is `no`, should we then implement the `id` scheme that we want before pushing this out?

File: src/status_im/chat/subs.cljs
Comment (mandrigin) [inline src/status_im/chat/subs.cljs]: is this a leftover?

File: src/status_im/chat/models/message.cljs
Comment (mandrigin) [inline src/status_im/chat/models/message.cljs:195]: do we call it with `false` anywhere?

File: src/status_im/data_store/realm/schemas/account/migrations.cljs
Comment (mandrigin) [inline src/status_im/data_store/realm/schemas/account/migrations.cljs]: was this done?

**Review by mandrigin — CHANGES_REQUESTED** [review]
a few questions after a first glance

File: src/status_im/chat/models/message.cljs
Comment (rasom) [reply src/status_im/chat/models/message.cljs:195]: yes, we do https://github.com/status-im/status-react/blob/812a2454aaa7d59f128b699016fdbb5c7aa103e7/src/status_im/chat/models/message.cljs#L346

File: src/status_im/data_store/realm/schemas/account/migrations.cljs
Comment (rasom) [reply src/status_im/data_store/realm/schemas/account/migrations.cljs]: no need to do this, the comment is removed

File: src/status_im/chat/subs.cljs
Comment (rasom) [reply src/status_im/chat/subs.cljs]: fixed

**Comment by rasom** [comment]
@cammellos 

> 1. If we include it, and then change id scheme, is there a clear path to upgrade id without breaking?

there is always a way to upgrade it without breaking, the question is how complicated it will be. If we change id scheme after merging this PR and before releasing `0.9.32` for those who upgrade from `0.9.31` upgrading will work the same way as without changing id schema (because it already has been changed after 0.9.31). Idk if that answers your question.

**Comment by cammellos** [comment]
> there is always a way to upgrade it without breaking, the question is how complicated it will be.

yes, seems like having an `old-old-message-id` (or something on those lines) might not be desirable, but necessary if we include this PR as it is, are we ok to pull the trigger?

**Comment by rasom** [comment]
@cammellos in overall:
there is no way to introduce change to message id schema and make functionality like our current implementation of replies (which relies on message id) compatible with the previous version of the app without adding some complexity like `very-very-old-message-id` and without adding necessary migrations for this.

**Review by cammellos — APPROVED** [review]

File: src/status_im/chat/models/input.cljs
Comment (mandrigin) [inline src/status_im/chat/models/input.cljs:90]: I would call them `message-id-v1` and `message-id-v2` here and everywhere.

**Review by mandrigin — CHANGES_REQUESTED** [review]

File: src/status_im/chat/models/input.cljs
Comment (rasom) [reply src/status_im/chat/models/input.cljs:90]: This PR will not be reviewable with such change and it will touch literally everything in the project with hundreds of possible regression. I really don't want to do this now.

Comment (mandrigin) [reply src/status_im/chat/models/input.cljs:90]: @rasom can you make a follow-up PR with just that? So we avoid this old-old thing?

Comment (rasom) [reply src/status_im/chat/models/input.cljs:90]: Do we really need this? I can make it, i just don't see how it will make things better.

Comment (mandrigin) [reply src/status_im/chat/models/input.cljs:90]: If we upgrade ID one more time, having no real naming convention will make a huge mess.

**Review by mandrigin — APPROVED** [review]

File: src/status_im/chat/models/input.cljs
Comment (rasom) [reply src/status_im/chat/models/input.cljs:90]: So that's what we need to do, introduce proper versioning mechanism with naming conventions, etc. We will not solve this right now by renaming `message-id` but will have huge chances to introduce a lot of other problems.

**Comment by rasom** [comment]
I will merge this PR a bit later if it's test and ready for merging @asemiankevich

File: src/status_im/chat/models/input.cljs
Comment (mandrigin) [reply src/status_im/chat/models/input.cljs:90]: yeah, I agree that this renaming/convention shouldn't be in the release

**Comment by asemiankevich** [comment]
@rasom it was tested before the last commit was done :)

**Comment by rasom** [comment]
@asemiankevich there was no "Tested OK" label, that's why i wanted to check

**Comment by Serhy** [comment]
Please note that autotests https://github.com/status-im/status-react/pull/6976/#issuecomment-443808447 were run not agains latest commit (build from https://github.com/status-im/status-react/pull/6976/#issuecomment-443717773 ), not the  https://github.com/status-im/status-react/pull/6976/commits/cb12e7e629157428dad74ffb7c3a4c3d2586975f

**Comment by rasom** [comment]
will be merged in #6987

**Pull request closed without merging**
